File: bucketbase/memory_bucket.py
from pathlib import PurePosixPath
from threading import RLock
from typing import Iterable, Union

from streamerate import slist, sset, stream

from bucketbase import DeleteError
from bucketbase.ibucket import ShallowListing, IBucket


class MemoryBucket(IBucket):
    """
    Implements IObjectStorage interface, but stores all objects in memory.
    This class is intended to be used for testing purposes only.
    """

    def __init__(self) -> None:
        self._objects = {}  # Store files
        self._lock = RLock()

    def put_object(self, object_name: PurePosixPath | str, content: Union[str, bytes, bytearray]) -> None:
        _object_name = self._validate_name(object_name)

        _content = self._encode_content(content)
        with self._lock:
            self._objects[_object_name] = _content

    def get_object_content(self, object_name: PurePosixPath) -> bytes:
        _object_name = self._validate_name(object_name)

        with self._lock:
            if _object_name not in self._objects:
                raise FileNotFoundError(f"Object {_object_name} not found in MemoryObjectStore")
            return self._objects[_object_name]

    def list_objects(self, prefix: PurePosixPath) -> slist[PurePosixPath]:
        self._split_prefix(prefix)  # validate prefix
        str_prefix = str(prefix)
        with self._lock:
            result = stream(self._objects).filter(lambda obj: str(obj).startswith(str_prefix)).map(PurePosixPath).to_list()
        return result

    def shallow_list_objects(self, prefix: PurePosixPath) -> ShallowListing:
        self._split_prefix(prefix)  # validate prefix
        str_prefix = str(prefix)
        pref_len = len(str_prefix)
        objects = slist()
        prefixes = sset()
        with self._lock:
            for sobj in self.list_objects(prefix).map(str):
                if "/" not in sobj[pref_len:]:
                    objects.append(PurePosixPath(sobj))
                else:
                    suffix = sobj[pref_len:]
                    common_suffix = suffix.split("/", 1)[0]
                    common_prefix = str_prefix + common_suffix + "/"
                    prefixes.add(common_prefix)
        return ShallowListing(objects=objects, prefixes=prefixes.to_list())

    def exists(self, object_name: PurePosixPath | str) -> bool:
        _object_name = self._validate_name(object_name)
        with self._lock:
            return _object_name in self._objects

    def remove_objects(self, list_of_objects: Iterable[PurePosixPath | str]) -> slist[DeleteError]:
        _list_of_objects = [str(obj) for obj in list_of_objects]

        delete_errors = slist()
        with self._lock:
            for obj in _list_of_objects:
                obj = self._validate_name(obj)
                if obj in self._objects:
                    self._objects.pop(obj)
        return delete_errors

File: java/src/main/java/com/esamtrade/bucketbase/MemoryBucket.java
package com.esamtrade.bucketbase;

import java.io.ByteArrayInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.*;
import java.util.concurrent.locks.ReentrantLock;
import java.util.stream.Collectors;

public class MemoryBucket extends BaseBucket {
    /**
     * Implements BaseBucket interface, but stores all objects in memory.
     * This class is intended to be used for testing purposes only.
     */

    private final Map<String, byte[]> objects;
    private final ReentrantLock lock;

    public MemoryBucket() {
        this.objects = new HashMap<>();
        this.lock = new ReentrantLock();
    }

    @Override
    public void putObject(PurePosixPath name, byte[] content) {
        String _name = validateName(name);
        byte[] _content = encodeContent(content);
        lock.lock();
        try {
            objects.put(_name, _content);
        } finally {
            lock.unlock();
        }
    }

    @Override
    public void putObjectStream(PurePosixPath name, InputStream stream) throws IOException {
        byte[] _content = stream.readAllBytes();
        putObject(name, _content);
    }

    @Override
    public byte[] getObject(PurePosixPath name) throws FileNotFoundException {
        String _name = validateName(name);
        lock.lock();
        try {
            if (!objects.containsKey(_name)) {
                throw new FileNotFoundException("Object " + _name + " not found in MemoryBucket");
            }
            return objects.get(_name);
        } finally {
            lock.unlock();
        }
    }

    @Override
    public ObjectStream getObjectStream(PurePosixPath name) throws FileNotFoundException {
        byte[] content = getObject(name);
        return new ObjectStream(new ByteArrayInputStream(content), name.toString());
    }

    @Override
    public List<PurePosixPath> listObjects(PurePosixPath prefix) {
        splitPrefix(prefix); // validate prefix
        String strPrefix = prefix.toString();
        lock.lock();
        try {
            return objects.keySet().stream()
                    .filter(obj -> obj.startsWith(strPrefix))
                    .map(PurePosixPath::from)
                    .collect(Collectors.toList());
        } finally {
            lock.unlock();
        }
    }

    @Override
    public ShallowListing shallowListObjects(PurePosixPath prefix) throws IOException {
        splitPrefix(prefix); // validate prefix
        String strPrefix = prefix.toString();
        int prefLen = strPrefix.length();
        List<PurePosixPath> objectsList = new ArrayList<>();
        Set<PurePosixPath> prefixesSet = new HashSet<>();
        lock.lock();
        try {
            for (PurePosixPath obj : listObjects(prefix)) {
                String sobj = obj.toString();
                if (!sobj.substring(prefLen).contains("/")) {
                    objectsList.add(obj);
                } else {
                    String suffix = sobj.substring(prefLen);
                    String commonSuffix = suffix.split("/", 2)[0];
                    String commonPrefix = strPrefix + commonSuffix + "/";
                    prefixesSet.add(PurePosixPath.from(commonPrefix));
                }
            }
        } finally {
            lock.unlock();
        }
        return new ShallowListing(objectsList, new ArrayList<>(prefixesSet));
    }

    @Override
    public boolean exists(PurePosixPath name) {
        String _name = validateName(name);
        lock.lock();
        try {
            return objects.containsKey(_name);
        } finally {
            lock.unlock();
        }
    }

    @Override
    public List<DeleteError> removeObjects(List<PurePosixPath> names) {
        List<PurePosixPath> _listOfObjects = names.stream().collect(Collectors.toList());
        List<DeleteError> deleteErrors = new ArrayList<>();
        lock.lock();
        try {
            for (PurePosixPath obj : _listOfObjects) {
                String validatedObj = validateName(obj);
                objects.remove(validatedObj);
            }
        } finally {
            lock.unlock();
        }
        return deleteErrors;
    }
}
